File: p.89/Python/solution.py
import math

def roman_to_int(roman):
    roman_values = {'I': 1, 
                    'V': 5, 
                    'X': 10, 
                    'L': 50, 
                    'C': 100, 
                    'D': 500, 
                    'M': 1000}
    value = 0
    total = 0
    prev_value = math.inf
    for char in roman:
        value = roman_values[char]
        if value > prev_value:
            total -= prev_value * 2
        prev_value = value
        total += value
    return total


def int_to_roman(num):
    int_to_roman_mappings = [
        (1000, 'M'), (900, 'CM'), (500, 'D'), (400, 'CD'),
        (100, 'C'), (90, 'XC'), (50, 'L'), (40, 'XL'),
        (10, 'X'), (9, 'IX'), (5, 'V'), (4, 'IV'), (1, 'I')
    ]
    result = []
    for value, numeral in int_to_roman_mappings:
        count = num // value
        result.append(numeral * count)
        num -= value * count
    return ''.join(result)

with open('roman.txt') as file:
    numeral = file.read().strip().splitlines()

savings = 0
for roman in numeral:
    original_length = len(roman)
    minimal_length = len(int_to_roman(roman_to_int(roman)))
    savings += original_length - minimal_length

print(f"Total characters is {savings}")

File: p.89/Kotlin/solution.kt
import java.io.File
import java.lang.Integer.MAX_VALUE

fun romanToInt(roman: String): Int {
  val romanValues = mapOf(
    'I' to 1, 'V' to 5, 'X' to 10, 'L' to 50,
    'C' to 100, 'D' to 500, 'M' to 1000
  )

  var total = 0
  var prevValue = MAX_VALUE

  for (char in roman) {
    val value = romanValues[char] ?: 0
    if (value > prevValue) {
      total += value - 2 * prevValue
    } else {
      total += value
    }
    prevValue = value
  }

  return total
}

fun intToMinimalRoman(num: Int): String {
  val mappings = listOf(
    1000 to "M", 900 to "CM", 500 to "D", 400 to "CD",
    100 to "C", 90 to "XC", 50 to "L", 40 to "XL",
    10 to "X", 9 to "IX", 5 to "V", 4 to "IV", 1 to "I"
  )

  val result = StringBuilder()
  var number = num

  for ((value, numeral) in mappings) {
    while (number >= value) {
      result.append(numeral)
      number -= value
    }
  }

  return result.toString()
}


fun calculateSavings(filename: String): Int {
  var savings = 0

  File(filename).forEachLine { line ->
    val roman = line.trim()
    val originalLength = roman.length
    val minimalLength = intToMinimalRoman(romanToInt(roman)).length
    savings += originalLength - minimalLength
  }

  return savings
}

fun main() {
  val filename = "roman.txt"
  val savings = calculateSavings(filename)
  println("Total characters saved : $savings")
}
